source/opt/loop_descriptor.cpp

```diff
@@ -0,0 +1,162 @@
+        // have a loop preheader.
+        return nullptr;
+      }
+      loop_pred = node->bb_;
+    }
+  }
+  // Safe guard against invalid code, SPIR-V spec forbids loop with the entry
+  // node as header.
+  assert(loop_pred && "The header node is the entry block ?");
+
+  // So we have a unique basic block that can enter this loop.
+  // If this loop is the unique successor of this block, then it is a loop
+  // preheader.
+  bool is_preheader = true;
+  uint32_t loop_header_id = loop_header_->id();
+  loop_pred->ForEachSuccessorLabel(
+      [&is_preheader, loop_header_id](const uint32_t id) {
+        if (id != loop_header_id) is_preheader = false;
+      });
+  if (is_preheader) return loop_pred;
+  return nullptr;
+}
+
+LoopDescriptor::LoopDescriptor(const Function* f) { PopulateList(f); }
+
+void LoopDescriptor::PopulateList(const Function* f) {
+  IRContext* context = f->GetParent()->context();
+
+  opt::DominatorAnalysis* dom_analysis =
+      context->GetDominatorAnalysis(f, *context->cfg());
+
+  loops_.clear();
+
+  // Traverse the tree and apply the above functor to find all the OpLoopMerge
+  // instructions. Instructions will be in domination order of BasicBlocks.
+  // However, this does not mean that dominance is implied by the order of
+  // loop_merge_inst you still need to check dominance between each block
+  // manually.
+  opt::DominatorTree& dom_tree = dom_analysis->GetDomTree();
+  // Post-order traversal of the dominator tree: inner loop will be inserted
+  // first.
+  for (opt::DominatorTreeNode& node :
+       ir::make_range(dom_tree.post_begin(), dom_tree.post_end())) {
+    Instruction* merge_inst = node.bb_->GetLoopMergeInst();
+    if (merge_inst) {
+      // The id of the continue basic block of this loop.
+      uint32_t merge_bb_id = merge_inst->GetSingleWordOperand(0);
+
+      // The id of the continue basic block of this loop.
+      uint32_t continue_bb_id = merge_inst->GetSingleWordOperand(1);
+
+      // The continue target of this loop.
+      BasicBlock* merge_bb = context->cfg()->block(merge_bb_id);
+
+      // The continue target of this loop.
+      BasicBlock* continue_bb = context->cfg()->block(continue_bb_id);
+
+      // The basic block containing the merge instruction.
+      BasicBlock* header_bb = context->get_instr_block(merge_inst);
+
+      // Add the loop to the list of all the loops in the function.
+      loops_.emplace_back(MakeUnique<Loop>(context, dom_analysis, header_bb,
+                                           continue_bb, merge_bb));
+      Loop* current_loop = loops_.back().get();
+
+      // We have a bottom-up construction, so if this loop has nested-loops,
+      // they are by construction at the tail of the loop list.
+      for (auto itr = loops_.rbegin() + 1; itr != loops_.rend(); ++itr) {
+        Loop* previous_loop = itr->get();
+
+        // If the loop already has a parent, then it has been processed.
+        if (previous_loop->HasParent()) break;
```

Comment:
I'm not convinced this is correct:

```
void foo() {
  for( ... ) { // loop 4
    for( ... ) {} // loop 1
    for( ... ) {  // loop 3
      for ( ... ) {} // loop 2
    }
  }
}
```

Think you could process in the order in which they are numbered.  Then in this case, loop 4 will not be set as the parent of loop 1.  If this is wrong, please explain what I have missed.

Reply:
Oh good catch.

We do a post-order df traversal for the dominator tree (from the main loop : `ir::make_range(dom_tree.post_begin(), dom_tree.post_end())`), so the inner most loops are visited first. So here loop 3 is dominated by loop 1, so loop 3 will be processed first.

But with this example:
```
void foo() {
  for( ... ) { // loop 4
    for( ... ) {  // loop 3
      for ( ... ) {} // loop 2
    }
    for( ... ) {} // loop 1
  }
}
```
The visit order will be 1, 2, 3, 4. When processing loop 3, loop 2 will have 3 as parent, and when processing loop 4, we will stop at loop 2 leaving loop 1 unset...